image: deprecate IDFromDigest(), and some minor fixes/cleanup

thaJeztah opened the pull request:
Having this function hides what it's doing, which is just to type-cast to an image.ID (which is a digest). Using a cast is more transparent, so deprecating this function in favor of a regular typecast.

I'm looking at potentially making this type a straight alias for digest.Digest (and possibly sunsetting it)

thaJeztah committed:
image: deprecate IDFromDigest()

Having this function hides what it's doing, which is just to type-cast
to an image.ID (which is a digest). Using a cast is more transparent,
so deprecating this function in favor of a regular typecast.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
image: store.restore(): use more structured logs

This code below is run when restoring all images (which can be "many"),
constructing the "logrus.WithFields" is deliberately not "DRY", as the
logger is only used for error-cases, and we don't want to do allocations
if we don't need it. A "f" type-alias was added to make it ever so slightly
more DRY, but that's just for convenience :)

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
daemon/images: imageKey() accept string

To reduce some type-juggling :)

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
image: store: rename vars that collided with types or shadowed vars

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

vvoland commented on image/store.go:
```diff
@@ -67,21 +67,21 @@ func NewImageStore(fs StoreBackend, lss LayerGetReleaser) (Store, error) {
 func (is *store) restore() error {
 	err := is.fs.Walk(func(dgst digest.Digest) error {
-		img, err := is.Get(IDFromDigest(dgst))
+		img, err := is.Get(ID(dgst))
```
Instead of repeating the .WithField() I tend  to first define partially filled log for the scope that the information is relevant
So, in this case, here I'd add:
```suggestion
		img, err := is.Get(IDFromDigest(dgst))
		log := logrus.WithField("digest", dgst)
```

thaJeztah replied:
Ah, yes, we do that in various places. It came up  when I was writing it, but not sure why I didn't do that.

Generally, I try to avoid using `log` for these (as it's easy to confuse for "this is the `log` package), so we use a slightly different name (e.g. `logger`)

vvoland commented on image/store.go:
```diff
@@ -67,21 +67,21 @@ func NewImageStore(fs StoreBackend, lss LayerGetReleaser) (Store, error) {
+		img, err := is.Get(ID(dgst))
 		if err != nil {
-			logrus.Errorf("invalid image %v, %v", dgst, err)
+			logrus.WithField("digest", dgst).WithError(err).Error("invalid image")
```
and then this line could be just:
```suggestion
			log.WithError(err).Error("invalid image")
```

vvoland commented on image/store.go:
```diff
@@ -67,21 +67,21 @@ func NewImageStore(fs StoreBackend, lss LayerGetReleaser) (Store, error) {
 			return nil
 		}
 		var l layer.Layer
 		if chainID := img.RootFS.ChainID(); chainID != "" {
```
here we can shadow previous log in this scope with a new `log` with chainID and os:
```suggestion
		if chainID := img.RootFS.ChainID(); chainID != "" {
			log := logrus.WithField("chainID", chainID).WithField("os", img.OperatingSystem())
```

vvoland commented on image/store.go:
```diff
@@ -67,21 +67,21 @@ func NewImageStore(fs StoreBackend, lss LayerGetReleaser) (Store, error) {
 		if chainID := img.RootFS.ChainID(); chainID != "" {
 			if !system.IsOSSupported(img.OperatingSystem()) {
-				logrus.Errorf("not restoring image with unsupported operating system %v, %v, %s", dgst, chainID, img.OperatingSystem())
+				logrus.WithField("digest", dgst).WithField("chainID", chainID).WithField("os", img.OperatingSystem()).Errorf("not restoring image with unsupported operating system")
```
this one can be reduced to:
```suggestion
				log.Error("not restoring image with unsupported operating system")
```

vvoland replied:
also, Errorf -> Error

vvoland commented on image/store.go:
```diff
@@ -67,21 +67,21 @@ func NewImageStore(fs StoreBackend, lss LayerGetReleaser) (Store, error) {
 			if err != nil {
 				if err == layer.ErrLayerDoesNotExist {
-					logrus.Errorf("layer does not exist, not restoring image %v, %v, %s", dgst, chainID, img.OperatingSystem())
+					logrus.WithField("digest", dgst).WithField("chainID", chainID).WithField("os", img.OperatingSystem()).Errorf("layer does not exist, not restoring image")
```
this one can be shortened too:
```suggestion
					log.Error("layer does not exist, not restoring image")
```

vvoland reviewed (commented):
+1 on using structured logs. I have some suggestions on DRYing the `WithField` calls.
Let me know what you think.

thaJeztah committed:
use errors.Is() to handle image store errors

The image store's used are an interface, so there's no guarantee
that implementations don't wrap the errors. Make sure to catch
such cases by using errors.Is.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
@vvoland updated; ptal 👍

vvoland reviewed (approved):
LGTM! Just a two nits 😄

thaJeztah commented on image/store.go:
```diff
@@ -67,21 +67,23 @@ func NewImageStore(fs StoreBackend, lss LayerGetReleaser) (Store, error) {
+			logger = logger.WithField("chainID", chainID).WithField("os", img.OperatingSystem())
 			if !system.IsOSSupported(img.OperatingSystem()) {
-				logrus.Errorf("not restoring image with unsupported operating system %v, %v, %s", dgst, chainID, img.OperatingSystem())
+				logger.Errorf("not restoring image with unsupported operating system")
```
Oh! Good catch

thaJeztah commented:
@corhere @neersighted ptal

corhere commented on image/store.go:
```diff
@@ -68,21 +68,23 @@ func NewImageStore(fs StoreBackend, lss LayerGetReleaser) (Store, error) {
 		}
 		var l layer.Layer
 		if chainID := img.RootFS.ChainID(); chainID != "" {
+			logger = logger.WithField("chainID", chainID).WithField("os", img.OperatingSystem())
```
Nit: chaining two `(*logrus.Entry).WithField` calls is surprisingly inefficient compared to `.WithFields` [as it is implemented in terms of the latter](https://github.com/sirupsen/logrus/blob/f8bf7650dccb756cea26edaf9217aab85500fe07/entry.go#L119-L122) and deep-copies the `Entry` on each call.

```suggestion
			logger = logger.WithFields(logrus.Fields("chainID": chainID, "os": img.OperatingSystem()))
```

<details><summary>Benchmarks</summary>

```
goos: darwin
goarch: amd64
pkg: github.com/corhere/logbench
cpu: Intel(R) Core(TM) i9-8950HK CPU @ 2.90GHz
BenchmarkWithField-12     	   10000	    344272 ns/op	  338640 B/op	      15 allocs/op
BenchmarkWithFields-12    	 1000000	      2372 ns/op	     566 B/op	       3 allocs/op
PASS
```

```go
package logbench

import (
	"io"
	"strconv"
	"testing"

	"github.com/sirupsen/logrus"
)

func init() {
	logrus.SetOutput(io.Discard)
}

func BenchmarkWithField(b *testing.B) {
	entry := logrus.NewEntry(logrus.StandardLogger())
	b.ResetTimer()
	b.ReportAllocs()
	for i := 0; i < b.N; i++ {
		entry = entry.WithField(strconv.Itoa(i), i)
	}
	entry.Info("hi")
}

func BenchmarkWithFields(b *testing.B) {
	entry := logrus.NewEntry(logrus.StandardLogger())
	fields := logrus.Fields{}
	b.ResetTimer()
	b.ReportAllocs()
	for i := 0; i < b.N; i++ {
		fields[strconv.Itoa(i)] = i
	}
	entry.WithFields(fields).Info("hi")
}
```

</details>

thaJeztah replied:
That's actually a good point here; this code is ran on `ImageStore.restore()`, which iterates over all images (which could be "many"). The `logger` was added to make the code a bit more DRY (see discussion with @vvoland above https://github.com/moby/moby/pull/44426#discussion_r1017921610), but looking now, I don't think that's a good idea _for this specific case_;

- the logger is only used in error conditions (the code is treats these as non-fatal, but logs them instead)
- but the `logger` would be constructed unconditionally, which means a lot of allocations for something that (in most cases) wouldn't be used.

Let me change it back to instantiate the `logger.WithField(s)` only if it's gonna be used. It's less "pretty" but probably good for this case.

neersighted reviewed (commented):
Looks like a slight typo in the commit message of [179ca12](https://github.com/moby/moby/pull/44426/commits/179ca1231685f78ee4bbf02a3bbe48199b0a2cc0) -- "wrapp"

Otherwise, LGTM.

thaJeztah commented:
Fixed the typo; let me bring this one in 👍
